### DISCUSSION TITLE
Identity Map does not work with Custom Type

### DISCUSSION THREAD

**Discussion by kangugau** [mikro-orm/mikro-orm#4335] [General]
**Is your feature request related to a problem? Please describe.**
If I define a entity with custom type primary key and find the same id twice, 2 SELECT query will be fired.

Example: 
`Author` has custom type `id`, the below script will call SELECT query 2 times
```js
import { v1 } from 'uuid';

class UuidBinaryType extends Type<string, Buffer> {
  convertToDatabaseValue(uuid) {
    return uuid && Buffer.from(parse(uuid));
  }

  convertToJSValue(bin) {
    if (!bin) return null;
    if (typeof bin === 'string') return bin;
    return stringify(bin);
  }

  getColumnType() {
    return 'binary(16)';
  }
}

class Author {
 @PrimaryKey({ type: UuidBinaryType })
  id!: string = v1();
}

const authorRepository = em.getRepository(Author);
const jon1 = await authorRepository.findOne("55fad9d0-e59e-11ed-a16a-c7f8809e6a6c");
const jon2 = await authorRepository.findOne("55fad9d0-e59e-11ed-a16a-c7f8809e6a6c");
```

**Describe the solution you'd like**
The query will be called only once

**Describe alternatives you've considered**
Maybe using database value instead of JS value as Identity Map's key

**Comment by B4nan** [comment]
The identity map _is_ using the database value. What version are you on?

**Reply by B4nan** [reply]
I see what you mean, weird the query is made - but the identity is kept, `jon1 === jon2` still holds even if there are two queries.

**Reply by kangugau** [reply]
I'm currently on v5.6.16

**Reply by kangugau** [reply]
I still think the problem is identity map using js value as key. The flow is:

1. lookup identity map (js value): **not found**
2. make a query
3. convert result to js value
4. add to identity map: **now it's found**  -> return the same one

**Reply by B4nan** [reply]
It's not entirely that simple, it's a mixture, some code paths use the converted value, and some of them use the raw one, and that is indeed wrong, the mixture is causing issues as converting the buffer to a string results in hex value without dashes, but if it goes through the type convertor they are there.

I am already looking into this.

**Reply by B4nan** [reply]
This should help:

https://github.com/mikro-orm/mikro-orm/commit/46b5e3ab50e8347fa6af0ca4289bf7d7ed7653d3
